python/rmm/rmm/statistics.py

```diff
@@ -0,0 +1,349 @@
+    """Context to enable allocation statistics.
+
+    If statistics has been enabled already (the current memory resource is an
+    instance of StatisticsResourceAdaptor), new counters are pushed on the
+    current allocation statistics stack when entering the context and popped
+    again when exiting using `push_statistics()` and `push_statistics()`.
+
+    If statistics has not been enabled, StatisticsResourceAdaptor is set as
+    the current RMM memory resource when entering the context and removed
+    again when exiting.
+
+    Raises
+    ------
+    ValueError
+        If the current RMM memory source was changed while in the context.
+    """
+
+    if push_statistics() is None:
+        # Save the current non-statistics memory resource for later cleanup
+        prior_non_stats_mr = rmm.mr.get_current_device_resource()
+        enable_statistics()
+    else:
+        prior_non_stats_mr = None
+
+    try:
+        current_mr = rmm.mr.get_current_device_resource()
+        yield
+    finally:
+        if current_mr is not rmm.mr.get_current_device_resource():
+            raise ValueError(
+                "RMM memory source stack was changed "
+                "while in the statistics context"
+            )
+        if prior_non_stats_mr is None:
+            pop_statistics()
+        else:
+            rmm.mr.set_current_device_resource(prior_non_stats_mr)
+
+
+class ProfilerRecords:
+    """Records of the memory statistics recorded by a profiler"""
+
+    @dataclass
+    class Data:
+        """Memory statistics of a single code block
+
+        Attributes
+        ----------
+        num_calls
+            Number of times this code block was evoked.
+        memory_total
+            Total number of bytes allocated
+        memory_peak
+            Peak number of bytes allocated
+        """
+
+        num_calls: int = 0
+        memory_total: int = 0
+        memory_peak: int = 0
+
+        def add(self, memory_total: int, memory_peak: int):
+            self.num_calls += 1
+            self.memory_total += memory_total
+            self.memory_peak = max(self.memory_peak, memory_peak)
+
+    def __init__(self) -> None:
+        self._lock = threading.Lock()
+        self._records: Dict[str, ProfilerRecords.Data] = defaultdict(
+            ProfilerRecords.Data
+        )
+
+    def add(self, name: str, data: Statistics) -> None:
+        """Add memory statistics to the record named `name`
+
+        This method is thread-safe.
+
+        Parameters
+        ----------
+        name
+            Name of the record
+        data
+            Memory statistics of `name`
+        """
+        with self._lock:
+            self._records[name].add(
+                memory_total=data.total_bytes, memory_peak=data.peak_bytes
+            )
+
+    @property
+    def records(self) -> Dict[str, Data]:
+        """Dictionary mapping record names to their memory statistics"""
+        return dict(self._records)
+
+    def pretty_print(
```

Comment:
I expect a method named `pretty_print` to actually do the printing. This just returns a string. We might want to rename this to something more generic like `ProfilerRecords.report()` to avoid implying that it prints anything to stdout.